### DISCUSSION TITLE
Custom Formatter Not Displaying Data

### DISCUSSION THREAD

**Discussion by kushfest** [anvilistas/tabulator#41] [Q&A] [answered]
I'm currently hitting a wall when trying to create a custom formatter. The column simply appears empty. The header is there, but all the cells below do not contain data.  I don't see any errors in the anvil console, so I dont know how to troubleshoot further.

My Tabulator component is set to load one of my app_tables. One of the columns in my table is called contact_types. It is a link to another table in which a series of columns are either set to True or None. So for my purposes, it is a dict of keys and values. I want to iterate through that list and for any key with the value of True, add an the name of the key and maybe a little icon.

I've tried this numerous ways using a form as the formatter, using a custom formatter that return a RichText item with the content set to a string of HTML code built by the formatter and when those didn't work, I've simplified to returning a string, but still no luck.  I have another custom formatter working fine, so I'm truly stumped

Here is my custom formatter code:

```
    def type_formatter(cell, **params):

      def clean_type(data):
        if len(data) == 2:
          return data.upper()
        else:
          return data.replace("_"," ").title()

      dic = dict(cell.get_value())
      content = ""
      for types in dic.items():
        if types[0] != 'name' and types[1] == True:
          con = clean_type(types[0]) + ", "
          content += con
      return content[:-2]
```

I create the column in my columns list with this code:
```
{"title": "Contact Type", "field": "contact_type", "formatter": type_formatter, "headerSort": False},
```

I'd be happy to send a clone link privately, but would rather not post it. Let me know if any other info would be helpful. Thanks in advance!

**Comment by s-cork** [comment, marked as answer]
first thing to try is to decorate the function with `@anvi.js.report_exceptions`
Any errors in python will be reported


What is probably happening though is that your formatter is calling the server when doing `dict(cell.get_value())`

This causes a Tabulator to receive a Promise object
But it doesn't know how to handle a Promise object
So it just renders nothing.
It probably doesn't throw an error because Js Tabulator just assumes that's what you intended to do.
Python likes to throw errors, JavaScript avoids it at all costs.

There's some documentation about blocking code and Promises in Anvil in the accessing Javascript section.
search for `blocking code` if you want to learn more.

I have quite an obscure suggestion

```python
def custom_getter(row, field):
    if field == "content_type":
        return dict(row["content_type"])
    return row[field]

...
    self.tabulator_1.options = {
       app_table...
       getter: custom_getter
    }
```

Why this might work - the getter happens in a place that can suspend and Javascript is expecting a Promise

**Reply by kushfest** [reply]
First off...Thank you! Truly appreciate your insights. I'm pretty quickly out of my depth, but determined to work this out. I think I get what you're saying about the Promise object and the need to wait for the promise to completed before we have the actual data. So I tried changing my ```dict(cell.get_value())``` command to ```dic = dict(anvil.js.await_promise(cell.get_value()))```. but that has so far not changed anything and the column is still empty. I can print the string my code creates, but i can't display it in a column yet.

I'm curious to dig more into your other suggestion as well re: the custom-getter. I'll ned to look at that with fresh eyes, but would i need to change the cel..get_value() command to utilize the custom getter? I'm a little lost how I use that.

**Reply by s-cork** [reply]
So the `custom_getter` overrides how we retrieve values from the row object.
By default the `getter` just returns `row['field']`.

You won't need to call `dict` on `cell.get_value()` because the `custom_getter` gave the cell a dict already.

_(The original use case for the `getter` option was for `Using Models` which you can find at the end of the README docs)_

**Reply by s-cork** [reply]
Also
You can't await the Promise  there because it's not a Promise in Python
It's only Promise in Javascript

Did you add the `getter` option already?

**Reply by kushfest** [reply]
It worked! You led the horse to water and now with some minor tweaks to adapt to my code., the column is no longer empty! That said,  I am still running ```dict``` on the ```cell.get_value()```. If I don't, I end up with a proxyobject, which doesn't seem to be a dict as i know them. Given how much time I spent trying to figure this out, I'll take the win haha, but are there any down sides to this approach?

**Reply by s-cork** [reply]
Not that I can see. You can access the original table row by calling `get_table_row()`. 

A proxy object is dictionary like so mostly works like a dict. 

Some features of Tabulator don't  work as well with the app table option and it might be better to give it a list of dicts. 

If you hit some places that feel like they're misbehaving that might be the reason. Just raise an issue if you notice anything. 

But it should work great for displaying data and sorting it. It depends on the use case.

**Reply by kushfest** [reply]
It has taken a couple days of noodling to figure it out, but I finally have gotten this all working with my data being pulled into a list of dicts rather than just accessing the app_table and WOW, the speed improvement is massive.  I know longer need to create use the custom getter to get this to work, which is a funny twist. 

This definitely seems like the best way to use Tabulator for data sourced from an anvil app_table.

**Comment by s-cork** [comment]
fyi - i've added support for throwing Suspension Errors in common 'gotcha' places in tabulator
This would have thrown an exception at 

```
dic = dict(cell.get_value())
``` 
see #43
